### ISSUE TITLE
aspnet5 and ADFS 2

### ISSUE THREAD

**Issue by airomero** [aspnet/Security#605]
Hello, is supported authentication with ADFS 2? I haven´t found any sample. I need to authenticate using dnxcore. If it´s not supported i think it´d be a must have since not everyone have adfs > 2 or Azure AD. Thanks!

**Comment by PinpointTownes** [comment]
ADFS 2 only supports WS-Federation, which is not currently implemented in ASP.NET 5.
I will be [eventually supported] (https://github.com/aspnet/Security/issues/43), but AFAIK, there's no timeline yet.

**Comment by airomero** [comment]
Thanks for the answer, i think that it´d be nice to have it before the RTM release since is a feature that the framework must have to live into the microsoft ecosystem. ADFS 2 was released years ago, yes, but the protocols it supports are standards so i think a lot of people would be very happy to have it. Thanks!!

**Comment by Eilon** [comment]
Yup, no WS-Fed planned for the current release, but we are looking at implementing it later.

**Comment by airomero** [comment]
Thanks. Do you have any idea when could it be implemented? RTM, during 2016, during 2017..... I'd like to have an idea in order to consider or not starting to migrate some projects to aspnet5 that depende on ADFS2. Thanks again.

**Comment by Eilon** [comment]
Hopefully this calendar year. We understand the importance of it.

I wonder if you could use the existing Katana WS-Fed middleware as auth middleware in ASP.NET 5... we do have the ability to hook up some Katana-era middleware in the pipeline. Auth middleware is a bit trickier though.

@Tratcher any idea if that has even a chance of working?

**Comment by Tratcher** [comment]
No the Katana and AspNet 5 security infrastructures are not inter-operable.

**Comment by airomero** [comment]
I understand, but consider that maybe some places can´t change their infrastructure (in our case migrate to a newer version of ADFS) because they want to use aspnet5. I want to take advantage of aspnet5, but it would be great not to have an overhead of work in our infrastructure. Or if you tell me that i have to use the latest version of iis to run aspnet5 it'd be also a dissapointment. It would be nice to have at least a basic functionality for login and logout. Thanks and i hope you understand mi comments.

**Comment by PinpointTownes** [comment]
@airomero another approach would be to use a tiny OpenID Connect/WS-Federation bridge: your ASP.NET 5 could then use the official OIDC middleware and you wouldn't have to wait for the WS-Federation middleware.

FWIW, that's something that should be easy to implement with the OWIN version of ASOS and Katana's WS-Federation middleware: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/tree/dev.

**Comment by airomero** [comment]
The investigation i made in aspnet5 has been using openid, i´ll give it a try to the bridge. But please consider this as it is very important to oficially support WS-Federation.

**Comment by Eilon** [comment]
@airomero duly noted that it is important.

**Comment by PinpointTownes** [comment]
@airomero both the ASP.NET team and the Azure AD team (that maintains IdentityModel) have limited resources and they had to cut many features from RTM to respect the deadline (e.g SignalR and WebPages) and WS-Federation is one of them.

The WS-Federation middleware itself will be trivial to implement, but having WS-Federation primitives/deserializers/metadata retrievers is a prerequisite: [sadly, all these things are not ready yet](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/tree/KDev/src/Microsoft.IdentityModel.Protocols.WsFederation) (not to mention SAML tokens support).

**Comment by airomero** [comment]
I hope WS-Federation will be ready this year since it is a prerequisite for us to implement aspnet5!!!!

**Comment by PinpointTownes** [comment]
Repeating it again and again doesn't really help :smile: 

If you absolutely need a WS-Federation middleware for your application, nothing prevents you from creating your own one.

**Comment by nickcam** [comment]
Hi, I've had a crack at creating some middleware to support Ws-Fed. Created a repository here: https://github.com/nickcam/Asp.Net5-WsFedAuth

It's pretty basic, essentially just inherits from / implements the cookie auth middleware, but adds some hooks and options to support WsFed. The main part is really just deciphering the SAML token returned from the IdP. 

Doesn't support dnxcore though (uses a bunch of 4.5.1 WIF classes) , and couldn't get the Katana Ws-Fed middleware working as per @Tratcher 's comment.

I'm not sure if there's a better way to do it, if anyone's got any feedback or suggestions that would be great!

**Comment by muziekklas** [comment]
This is a total blocker for us to migrate our apps to .Net Core. Any timeline yet on release?
Thanks!

**Comment by Anton987** [comment]
Any news on this?

**Comment by Tratcher** [comment]
No news. Everything is blocked on this dependency: https://github.com/dotnet/corefx/issues/4278. Feel free to prod them.

**Comment by danroth27** [comment]
This looks like a dup of #43

**Comment by Eilon** [comment]
Yup, dup of https://github.com/aspnet/Security/issues/43.
